### PR TITLE
Csf-tools: Add `satisfies` support to ConfigFile

### LINKED ISSUE #20935: [Bug]: @storybook/csf-tools doesn't parse .storybook/main.ts config properly if using `satisfies StorybookConfig`
### Describe the bug

# Issue

If using typescript storybook config and [`satisfies` from typescript 4.9 ](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-9.html#the-satisfies-operator) to typecheck
```typescript
export default {
  // config here
} satisfies StorybookConfig;
```
instead of something like
```typescript
const config: StorybookConfig = {
  // config here
};
export default config;
```
the ast isn't handled properly.

As you can see below in the asts under "Additional context", when using the `satisfies` operator, the `ObjectExpression` is wrapped in any extra `TSSatisfiesExpression` node which causes the code to break.

A new unit test is probably needed in https://github.com/storybookjs/storybook/blob/next/code/lib/csf-tools/src/ConfigFile.test.ts that tests the `satisfies` operator.

### My storybook config for example:

.storybook/main.ts
```typescript
import type { StorybookConfig } from '@storybook/react-vite';

export default {
  stories: [],
  addons: [
    '@storybook/addon-essentials',
  ],
  framework: {
    name: '@storybook/react-vite',
    options: {},
  },
} satisfies StorybookConfig; // issue here
```

# Root Cause

This began in version `v7.0.0-beta.39`.

The following call seems to spam the warning log from the following line starting in `v7.0.0-beta.39`.
https://github.com/storybookjs/storybook/blob/v7.0.0-beta.39/code/lib/csf-tools/src/ConfigFile.ts#L155
```typescript
      ExportDefaultDeclaration: {
        enter({ node, parent }) {
          self.hasDefaultExport = true;
          const decl =
            t.isIdentifier(node.declaration) && t.isProgram(parent)
              ? _findVarInitialization(node.declaration.name, parent)
              : node.declaration;

          if (t.isObjectExpression(decl)) {
            self._exportsObject = decl;
            decl.properties.forEach((p: t.ObjectProperty) => {
              const exportName = propKey(p);
              if (exportName) {
                let exportVal = p.value;
                if (t.isIdentifier(exportVal)) {
                  exportVal = _findVarInitialization(exportVal.name, parent as t.Program);
                }
                self._exports[exportName] = exportVal as t.Expression;
              }
            });
          } else {
            logger.warn(`Unexpected ${JSON.stringify(node)}`); // spam logging here
          }
        },
      },
```

This is probably because of a newly added call to `readConfig` here that ends up hitting the previous piece of code, which didn't exist in `v7.0.0-beta.38`.
https://github.com/storybookjs/storybook/blob/v7.0.0-beta.39/code/lib/core-common/src/utils/validate-configuration-files.ts#L28
```typescript
    const main = await readConfig(mainConfigPath);
```

# Temporary workaround (if using pnpm)

Add to package.json
```json
{
  "pnpm": {
    "overrides": {
      "@storybook/core-common": "7.0.0-beta.38"
    }
  }
}
```

### To Reproduce

Can probably reproduce just by adding a new test to https://github.com/storybookjs/storybook/blob/next/code/lib/csf-tools/src/ConfigFile.test.ts that uses the satisfies operator`.

### System

```shell
Environment Info:

  System:
    OS: macOS 13.1
    CPU: (8) x64 Intel(R) Core(TM) i5-1038NG7 CPU @ 2.00GHz
  Binaries:
    Node: 19.5.0 - /usr/local/bin/node
    npm: 9.3.1 - /usr/local/bin/npm
  Browsers:
    Chrome: 109.0.5414.119
    Firefox: 90.0.2
    Safari: 16.2
```

### Additional context

# Unsupported AST (with `satisfies`)

```
Node {
  type: 'TSSatisfiesExpression',
  start: 78,
  end: 467,
  loc: null,
  expression: Node {
    type: 'ObjectExpression',
    start: 78,
    end: 441,
    loc: SourceLocation {
      start: [Position],
      end: [Position],
      filename: undefined,
      identifierName: undefined,
      lines: [Lines],
      tokens: [Array],
      indent: 0
    },
    properties: [ [Node], [Node], [Node], [Node] ],
    extra: { trailingComma: 438 }
  },
  typeAnnotation: Node {
    type: 'TSTypeReference',
    start: 452,
    end: 467,
    loc: SourceLocation {
      start: [Position],
      end: [Position],
      filename: undefined,
      identifierName: undefined,
      lines: [Lines],
      tokens: [Array],
      indent: 0
    },
    typeName: Node {
      type: 'Identifier',
      start: 452,
      end: 467,
      loc: [SourceLocation],
      name: 'StorybookConfig'
    }
  }
}
```

# Supported AST (without `satisfies`)

```
Node {
  type: 'ObjectExpression',
  start: 78,
  end: 441,
  loc: null,
  properties: [
    Node {
      type: 'ObjectProperty',
      start: 82,
      end: 212,
      loc: [SourceLocation],
      method: false,
      key: [Node],
      computed: false,
      shorthand: false,
      value: [Node]
    },
    Node {
      type: 'ObjectProperty',
      start: 216,
      end: 330,
      loc: [SourceLocation],
      method: false,
      key: [Node],
      computed: false,
      shorthand: false,
      value: [Node]
    },
    Node {
      type: 'ObjectProperty',
      start: 334,
      end: 402,
      loc: [SourceLocation],
      method: false,
      key: [Node],
      computed: false,
      shorthand: false,
      value: [Node]
    },
    Node {
      type: 'ObjectProperty',
      start: 406,
      end: 438,
      loc: [SourceLocation],
      method: false,
      key: [Node],
      computed: false,
      shorthand: false,
      value: [Node]
    }
  ],
  extra: { trailingComma: 438 }
}
```

### PR DESCRIPTION
**Pull request by zhyd1997**
Closes #20935 

## What I did

added `satisfies` support to `.storybook/main.ts`:
## Checklist

- [x] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [ ] Make sure to add/update documentation regarding your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [ ] If this PR should be tested against many or all sandboxes,
      make sure to add the `ci:merged` or `ci:daily` GH label to it.
- [ ] Make sure this PR contains **one** of the labels below.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

### CODE REVIEW

File: code/lib/csf-tools/src/ConfigFile.ts [hunk 1]
```diff
@@ -139,9 +139,12 @@ export class ConfigFile {
...
 
-          if (t.isObjectExpression(decl)) {
-            self._exportsObject = decl;
-            decl.properties.forEach((p: t.ObjectProperty) => {
+          if (t.isObjectExpression(decl) || t.isTSSatisfiesExpression(decl)) {
+            const expression = t.isTSSatisfiesExpression(decl)
+              ? (decl.expression as Extract<t.Expression, t.ObjectExpression>)
+              : decl;
+            self._exportsObject = expression;
+            expression.properties.forEach((p: t.ObjectProperty) => {
```
Comment (zhyd1997) [inline code/lib/csf-tools/src/ConfigFile.ts:147]: AST comparison:
Screenshot 2023-02-05 at 18 35 03
Screenshot 2023-02-05 at 18 36 25

**Review by shilman — APPROVED** [review]
Nice one @zhyd1997 !! Thanks for this 🙏

**Comment by zhyd1997** [comment]
@shilman 

I am not familiar with the codebase, so please review the PR carefully to make sure it will not break something, If you have any questions , please let me know.

File: code/lib/csf-tools/src/ConfigFile.ts [hunk 2]
```diff
@@ -139,9 +139,12 @@ export class ConfigFile {
               ? _findVarInitialization(node.declaration.name, parent)
               : node.declaration;
 
-          if (t.isObjectExpression(decl)) {
-            self._exportsObject = decl;
-            decl.properties.forEach((p: t.ObjectProperty) => {
+          if (t.isObjectExpression(decl) || t.isTSSatisfiesExpression(decl)) {
+            const expression = t.isTSSatisfiesExpression(decl)
+              ? (decl.expression as Extract<t.Expression, t.ObjectExpression>)
```
Comment (shilman) [inline code/lib/csf-tools/src/ConfigFile.ts:144]: I'm suspicious of this line. I think it should be something like:

```ts
          let decl =
            t.isIdentifier(node.declaration) && t.isProgram(parent)
              ? _findVarInitialization(node.declaration.name, parent)
              : node.declaration;
          if(t.isAsExpression(decl) || t.isTSSatisfiesExpression(decl)) {
            decl = decl.expression;
          }
          if(t.isObjectExpression(decl)) {
            ... 
          } else {
            logger.warn(`Unexpected ${JSON.stringify(node)}`);
          }
 ```
 
The problem with the current code is that it assumes the satisfies expression is an ObjectExpression, which might not be the case. For example:

```ts
export default 'foo' satisfies Bar;
```

**Review by shilman — CHANGES_REQUESTED** [review]
I didn't review closely enough. LMK what you think about this

**Comment by stevezhu** [comment]
Nice! Thanks for implementing this!

**Review by yannbf — APPROVED** [review]
Looks great to me, thanks @zhyd1997 !!

**Review by shilman — APPROVED** [review]

**Pull request merged**